### ISSUE TITLE
RFE: team_number does not yet support the optional team argument

### ISSUE THREAD

**Issue by MichaelSiehl** [sourceryinstitute/OpenCoarrays#524]
[issue stat img]: https://img.shields.io/issuestats/i/github/sourceryinstitute/OpenCoarrays.svg?style=flat-square

| Avg response time              |
|:------------------------------:|
| ![Issue Stats][issue stat img] |

# Request for Enhancement (RFE) #

- OpenCoarrays Version: `2.0.0`
- Fortran Compiler: `GNU Fortran (GCC) 8.0.1 20180225 (experimental)`
- C compiler used for building lib: `gcc (GCC) 8.0.1 20180225 (experimental)`
- Installation method: `install.sh`
- Output of `uname -a`: `Linux 3.13.0-144-generic #193-Ubuntu SMP Thu Mar 15 17:03:53 UTC 2018 x86_64 x86_64 x86_64 GNU/Linux`
- MPI library being used: `MPICH 3.2.1`
- Machine architecture and number of physical cores: `laptop computer, 4 cores`
- Version of CMake: `3.4.0`

## Observed Behavior ##

With OpenCoarrays 2.0.0, the Fortran 2018 intrinsic function team_number does not yet support the optional team argument. This should be a known issue because the runtime does give an appropriate runtime error message: `Fortran runtime error on image 1: team_number does not yet support the optional team argument`. Example program:
```fortran
program Main
! please compile and run with 15 coarray images
  !
  use,intrinsic :: ISO_FORTRAN_ENV, only: team_type
  implicit none
  !
  integer :: intTeamNumber
  type (team_type) :: MainTeam
  type (team_type) :: SubTeam
  !
!
!************************************************
! MainTeam **************************************
!************************************************
!
! split the 15 images of the initial team into 3 child teams consisting of 5 images each:
  if (this_image() <= 5) then
    intTeamNumber = 1
  else if (this_image() >= 6 .and. this_image() <= 10) then
    intTeamNumber = 2
  else if (this_image() >= 11) then
    intTeamNumber = 3
  end if
  !
form team (intTeamNumber, MainTeam) ! this creates the 3 child teams in MainTeam
!
change team (MainTeam)
!
! important: the following gets executed on the images of all the 3 newly created child teams:
!
!************************************************
! SubTeam ***************************************
!************************************************
!
! split the 5 images of the MainTeam's child teams into two distinct teams resp.:
  if (this_image() < 3) then
    intTeamNumber = 1
  else
    intTeamNumber = 2
  end if
!
form team (intTeamNumber, SubTeam) ! this generates 6 teams (2 new sub teams in the 3 main teams resp.)
                                   ! within the change team (MainTeam) construct
change team (SubTeam)
!
SubTeam_select: select case (team_number())
!
case (1) SubTeam_select
  ! codes for child team 1 of SubTeam:
  write(*,*) 'image number: ', this_image(), ' of team number: ', team_number(), &
             ' in SubTeam, of MainTeam''s child team number:', team_number(MainTeam)
              !Fortran runtime error on image 1: team_number does not yet support the optional team argument
  !
!*******************
case (2) SubTeam_select
  ! codes for child team 2 of SubTeam:
  write(*,*) 'image number: ', this_image(), ' of team number: ', team_number(), &
             ' in SubTeam, of MainTeam''s child team number:', team_number(MainTeam)
              !Fortran runtime error on image 1: team_number does not yet support the optional team argument
!
end select SubTeam_select
!
end team !(SubTeam)
!
!*************
!*************
!*************
!
end team !(MainTeam)
!
end program Main
```

## Expected Behavior ##

Just to give an indication of the importance of the optional team argument:
As the above example program can already show (it does not fail if compiled with team_number not using the optional team argument), if we execute the same form team statement on all images within a change team construct (that comprises a set of teams), that single form team statement does not only form one set of new teams but instead can form several sets of new teams: For each team in MainTeam a new set of teams is created in SubTeam; Since MainTeam comprises 3 teams, and the form team statement in SubTeam does form 2 new teams, that single form team statement does actually form 6 new teams.
With such nested teams, the team argument should allow to use the team_number intrinsic to get information about the nesting and the belonging of the images of a nested team with respect to their parent team. This could be helpful for the development of new kinds of parallel algorithms later on.

**Comment by rouson** [comment]
@MichaelSiehl Thanks for submitting this issue and especially for the code example. If you have any interest, the best way to get this feature added is to dive in.  Going forward, OpenCoarrays is going to be mostly a community effort with only small amounts of related project funding.  In case it helps, I have long considered the possibility of replacing some of the C code in OpenCoarrays with Fortran code as a way to increase the likelihood that the Fortran community will contribute code to OpenCoarrays.  As a simple example of what this would look like, take a look at the partial implementation of [`this_image()`] in Fortran inside the [opencoarrays module]:
```fortran
  ! Return the image number (MPI rank + 1)
  function this_image()  result(image_num)
#ifdef MPI_WORKING_MODULE
    use mpi, only : MPI_Comm_rank
#endif
    integer(c_int) :: image_num,ierr
   !image_num = opencoarrays_this_image(unused)
    call MPI_Comm_rank(CAF_COMM_WORLD,image_num,ierr)
    if (ierr/=0) call error_stop
    image_num = image_num + 1
  end function
```
The commented line makes a call to the C implementation of `this_image()`.  It's commented because I had originally planned for the above procedure to wrap C, but had trouble getting that to work and so decided that the function is suffiiciently simple that I just wrote it in Fortran rather than wrapping it.  A similar approach could be tried with `team_number()`.  If we go this route and also give the procedure the `bind(C)` attribute, then some of the C procedures could eventually be replaced with Fortran procedures without any impact on user code.  Would this make it more likely that you would consider implementing the requested feature?


[opencoarrays module]: https://github.com/sourceryinstitute/OpenCoarrays/blob/master/src/extensions/opencoarrays.F90

[`this_image()`]: https://github.com/sourceryinstitute/OpenCoarrays/blob/master/src/extensions/opencoarrays.F90#L760

**Comment by MichaelSiehl** [comment]
Hi Damian,

the ability to use Fortran for implementing new features sounds great and
yes, it could make it more likely to try implementing features by myself
(without any guaranty of success yet ;-). Still, the team_number intrinsic
function (as well as num_images, which does not support team variables at
compile time yet) is not on top of my wishing list of missing features. (We
can circumvent the missing optional team argument easily by using a local
variable to store team values – since using any of the team-related
statements has no impact to the values of local variables). There are
probably other missing features (e.g. allocatable derived type coarrays at
the team level) that I would consider to be more important yet, but still
NO single missing feature that does prevent from highly sophisticated use
of the Fortran 2018 team features with OpenCoarrays yet.

Further, and considering that upcoming ifort 19 (beta yet) will not already
support Fortran 2018 team features, I do suppose that OpenCoarrays will
continue stay way ahead for quite a long time.

Furthermore, right now I am full of enthusiasm about the working team
features and will prefer to focus on the working features and do further
research using them. I did already succeed with encapsulating F18 team
features into a (yet simple) customized create_team procedure (similar to a
customized synchronization procedure). At this early stage of testing, I
can not fully tell what the opportunities of this will be, but do still
believe that we can easily increase the possibilities of the basic (but
powerful) team features of Fortran 2018 using current OpenCoarrays already.

If things work out well, I will give working example codes in the coming
weeks and month.

My Best Regards


2018-04-25 23:32 GMT+02:00 Damian Rouson <notifications@github.com>:

**Comment by nathanweeks** [comment]
I've attempted a fix in https://github.com/sourceryinstitute/OpenCoarrays/pull/631. The resulting output of @MichaelSiehl 's code is:

```
$ cafrun -np 12 ./a.out
 image number:            1  of team number:            1  in SubTeam, of MainTeam's child team number:           1
 image number:            2  of team number:            1  in SubTeam, of MainTeam's child team number:           1
 image number:            1  of team number:            2  in SubTeam, of MainTeam's child team number:           1
 image number:            2  of team number:            2  in SubTeam, of MainTeam's child team number:           1
 image number:            3  of team number:            2  in SubTeam, of MainTeam's child team number:           1
 image number:            1  of team number:            1  in SubTeam, of MainTeam's child team number:           2
 image number:            2  of team number:            1  in SubTeam, of MainTeam's child team number:           2
 image number:            1  of team number:            2  in SubTeam, of MainTeam's child team number:           2
 image number:            2  of team number:            2  in SubTeam, of MainTeam's child team number:           2
 image number:            3  of team number:            2  in SubTeam, of MainTeam's child team number:           2
 image number:            1  of team number:            1  in SubTeam, of MainTeam's child team number:           3
 image number:            2  of team number:            1  in SubTeam, of MainTeam's child team number:           3
```

**Comment by MichaelSiehl** [comment]
@nathanweeks: Great job! Thanks very much.

**Comment by zbeekman** [comment]
Fixed in #631 and in release 2.5.0. Thanks @nathanweeks for implementing the fix!

And thanks to @MichaelSiehl for reporting the bug!
